added live build status to Telescope dashboard:

mqnguyen5 opened the pull request:
<!--
Thanks for sending a pull request!
- if this is your first time, please ensure you have read through our contributor guide: https://github.com/Seneca-CDOT/telescope/blob/master/docs/CONTRIBUTING.md
-->

## Issue This PR Addresses

Fixes #2416 

<!--
1. Automatically close the issue when this PR is merged
    USAGE: Fixes #<issue number>
2. If your PR addresses an issue but does not close it
    USAGE: #<issue number> <reason>(i.e. This issue was worked on by @user and myself and his PR should close the issue.)
-->

## Type of Change

<!-- bug fix, feature, documentation, UI, etc. -->

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

<!-- Please add a detailed description of what this PR does and why it is needed -->
- Added scripts for fetching and rendering build-header
- Added html template for build-header
- Imported build-header to run on load

<img width="1422" alt="Screen Shot 2021-12-08 at 1 01 37 PM" src="https://user-images.githubusercontent.com/59778930/145264347-b7c2770d-b8a2-4ec0-8487-a01fb38717e9.png">
<img width="1422" alt="Screen Shot 2021-12-08 at 1 22 59 PM" src="https://user-images.githubusercontent.com/59778930/145264349-415997e1-22e6-4296-9c2a-150ad258a707.png">
<img width="1422" alt="Screen Shot 2021-12-08 at 1 29 28 PM" src="https://user-images.githubusercontent.com/59778930/145264351-57fc5616-5158-458f-ab15-a5718c7f8e7b.png">

To test the feature locally, do the following:

- Change `autodeploymentUrl` in `public/js/buildl-log/api.js`

```js
const autodeploymentUrl = (path) => `//dev.api.telescope.cdot.systems:4000/${path}`;  
```

- Allow all `connectSrc` in `src/server.js`

```js
connectSrc: ["'self'", "*"]
```

- Change `current` to `previous` inside `src/api/status/public/js/build-log/api.js`

```js
export const checkBuildStatus = async () => {
  try {
    const res = await fetch(autodeploymentUrl('status'));
    if (!res.ok) {
      throw new Error('unable to get build info');
    }
    const data = await res.json();
    if (!data.current) {
      return { building: false };
    }
    return {
      building: true,
      title: data.type,
      githubData: data.current.githubData,
      startedAt: new Date(data.current.startedDate),
      stoppedAt: new Date(data.current.stopDate),
      result: data.code,
    };
  } catch (err) {
    console.error(err);
    return { building: false };
  }
};
```

## Checklist

<!-- Before submitting a PR, address each item -->

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

Andrewnt219 commented:
Can you update the branch and add an instruction for testing this feature?

Andrewnt219 commented on src/api/status/public/js/build-log/build-header.js:
```diff
@@ -0,0 +1,47 @@
+  buildDuration.innerText = `${minutes}m ${seconds}s`;
+}
+
+export default async function buildHeader() {
+  try {
+    const res = await fetch(autodeploymentUrl('status'));
+    if (!res.ok) {
+      const icon = document.createElement('i');
+      icon.className = 'fas fa-server px-2';
+      resMsg.innerHTML = '';
+      resMsg.append(icon);
+      resMsg.innerHTML += 'Unable to get build info.';
+
+      throw new Error('Unable to get build info');
+    }
+    const { previous } = await res.json();
+    renderBuildInfo(previous);
+  } catch (err) {
+    console.error(err);
+  }
+}
```
We don't need to double fetch from "/status". The `checkBuildStatus` is already called every 5 seconds. You can just pass down the `githubData` and call `renderBuildInfo` inside `checkForBuild` when there is an ongoing build.

Andrewnt219 replied:
That'll also keep the build-log and build-header in sync.

mqnguyen5 replied:
Previously, I tried to follow #2486 to test if the data was fetch successfully, but now I'm getting this error:

<img width="902" alt="Screen Shot 2021-12-08 at 2 00 40 PM" src="https://user-images.githubusercontent.com/59778930/145267874-5f7bcb3a-0f75-4ef8-80d9-9954d99aec28.png">

Andrewnt219 replied:
Content Security Error, you are missing this change.  But this is only for testing, do not include it in the commit.
```diff
-         connectSrc: [
-          "'self'",
-          '*.fontawesome.com',
-          `${host.replace(/(^\w+:|^)\/\//, '')}:4000`,
-          '*.github.com',
-        ],
+ connectSrc: ["'self'", "*"]

```

Andrewnt219 commented:
@humphd Since there's no way to run the autodeployment locally, maybe it's easier to point this to the stagging or prod autodployment? 

https://github.com/Seneca-CDOT/telescope/blob/38d74beb90624d16103287e5e44db1b22f932595/src/api/status/public/js/build-log/api.js#L2

humphd commented:
@Andrewnt219 great suggestion.  For reviewing this, that works, and we can change it back before we merge.  @mqnguyen5 can you point it at https://dev.telescope.cdot.systems for now?

mqnguyen5 commented:
Will do

Andrewnt219 commented on src/api/status/public/js/build-log/build-header.js:
```diff
@@ -0,0 +1,43 @@
+    }
+    renderBuildInfo(data);
+  } catch (err) {
+    console.error(err);
```
No need for a trycatch here.

Andrewnt219 commented on src/api/status/public/js/build-log/build-header.js:
```diff
@@ -0,0 +1,43 @@
+const buildDuration = document.getElementById('build-duration');
+
+function renderBuildInfo({ githubData, startedAt, stoppedAt, result }) {
+  resMsg.toggleAttribute('hidden');
+  buildHeaderInfo.toggleAttribute('hidden');
```
This makes the build header switch between loading state and info state every 5s. Like, 5s of info and 5s of loading.

mqnguyen5 replied:
You have any suggestions on how to fix it? I've tried using `setAttribute` and `removeAttribute` here but it doesn't seems to do the trick

Andrewnt219 replied:
Maybe the function should only render the build info, not toggling between two states (i.e. only render info state)? I don't think there's a case where toggling between those two is necessary.

Yoda-Canada commented:
I have changed ` current ` to ` previous ` inside src/api/status/public/js/build-log/api.js, but can't get the result which have the data. It only display loading status. How to get the previous information?
![image](https://user-images.githubusercontent.com/55161917/145304188-57b7d611-d06d-4b53-816c-38c53aa10f72.png)

![image](https://user-images.githubusercontent.com/55161917/145303627-57b5f5dc-6239-446d-9503-a8536af23d1c.png)

mqnguyen5 commented:
change inside the `if` statement as well

manekenpix commented:
We really need to get better at rebasing instead of merging master into our branches

![image](https://user-images.githubusercontent.com/23108901/145304850-2781b071-8518-4f45-b6e2-a163417f0f2b.png)

Yoda-Canada commented:
I changed, but the result is same.
![image](https://user-images.githubusercontent.com/55161917/145304915-fb716aa5-0502-4f09-827b-7ce6e9ba52d2.png)

mqnguyen5 commented:
@Yoda-Canada sorry, I just updated the instruction to test this feature. You mind trying it again?

mqnguyen5 commented:
> We really need to get better at rebasing instead of merging master into our branches
> 
> ![image](https://user-images.githubusercontent.com/23108901/145304850-2781b071-8518-4f45-b6e2-a163417f0f2b.png)

Sorry about that, you mind showing me how to do it?

Andrewnt219 commented on src/api/status/public/js/build-log/check-for-build.js:
```diff
@@ -42,4 +43,5 @@ export default async function checkForBuild() {
       reader.read().then(processLog).catch(finish);
     }
   }
+  buildHeader(status);
```
This needs to move up before the `if(build)`, so even when there is an ongoing build, the header will still be updated.

Andrewnt219 commented:
> you mind showing me a better way to do it instead of git pull upstream master?

Use `git pull --rebase upstream master`

Yoda-Canada commented:
@mqnguyen5 I followed your new steps, and it only displayed the loading page.

Andrewnt219 commented:
@mqnguyen5 Handlebars PR is merged, can you rebase this? The changes in build.html go into builds.hbs

mqnguyen5 commented:
Will do, I'll try to get it done soon.

Andrewnt219 commented:
> @mqnguyen5 I followed your new steps, and it only displayed the loading page. Do you use a specific address to get the following data?

@Yoda-Canada Have you changed all the `current` to `previous`?

Andrewnt219 commented:
@Yoda-Canada  can you show me the console on the build page? And you can reply in a new comment, editing my comment looks weird.

Yoda-Canada commented:
@Andrewnt219 
![image](https://user-images.githubusercontent.com/55161917/145439016-b1bb4f4c-1663-42d4-a3a2-4af36af19b32.png)

Andrewnt219 commented:
There's no `autodeploymentUrl` in build-header.js, you need to pull the new changes.

Yoda-Canada commented:
Ever though I have allowed  ` connectSrc: ["'self'", "*"] `, still can't fetch the data
![1639097849(1)](https://user-images.githubusercontent.com/55161917/145499476-d0140d97-7823-4c3b-bc13-55865c90da0e.png) 
![1639098192(1)](https://user-images.githubusercontent.com/55161917/145499648-55f67c13-639a-4547-a430-e16a3e9c5fce.png)

humphd commented:
When you say you added it, the error above shows `localhost:4000` but you're fetching from `http://dev.api.telescope.cdot.systems:4000/status` so it's not added.

Kevan-Y commented:
> I followed the instruction and changed everything. I think the code has an issue, right?

Look okay on my end.

Kevan-Y reviewed (commented):
![image](https://user-images.githubusercontent.com/58233223/145503326-2ade2441-4fb4-4b69-acd0-4bbb0da1ce6d.png)
Would it be better to have a bit more space on the left?

Andrewnt219 commented on src/api/status/public/js/build-log/api.js:
```diff
@@ -15,7 +16,10 @@ export const checkBuildStatus = async () => {
       title: data.type,
+      githubData: data.current.githubData,
       startedAt: new Date(data.current.startedDate),
+      stoppedAt: new Date(data.current.stopDate),
```
There's no `stopDate` for current, use the current date instead.

Andrewnt219 commented on src/api/status/public/js/build-log/build-header.js:
```diff
@@ -0,0 +1,39 @@
+const buildDuration = document.getElementById('build-duration');
+
+function renderBuildInfo({ githubData, startedAt, stoppedAt, result }) {
+  buildHeaderTitle.innerHTML = '';
+  buildHeaderTitle.innerText = 'Build Status';
```
overwriting `innerHtml` should be enough
```suggestion
  buildHeaderTitle.innerHTML = 'Build Status';
```

Andrewnt219 commented on src/api/status/src/views/builds.hbs:
```diff
@@ -32,6 +32,43 @@
     </div>
   </nav>
   <div class="container-fluid py-2">
+    <div id="build-header" class="border rounded p-2 mb-3 bg-white">
```
Can you make this one a [card](https://www.creative-tim.com/learning-lab/bootstrap/cards/material-dashboard)?

Andrewnt219 commented on src/api/status/src/views/builds.hbs:
```diff
@@ -32,6 +32,43 @@
+                <span id="build-sender-name"></span>
+              </a>
+              synchronized
+              <span><a id="build-git-sha" href="#" class="text-decoration-underline"></a></span>
```
This can be `card-link`

Yoda-Canada commented on src/api/status/public/js/build-log/build-header.js:
```diff
@@ -0,0 +1,39 @@
+  buildSenderImg.src = githubData.sender.avatar_url;
+  buildGitSHA.setAttribute('href', githubData.compare);
+  buildGitSHA.innerText = githubData.after.substring(0, 7);
+  buildResult.innerText = result === 0 ? 'Good' : 'Error';
```
According to the data ,  ``` result: data.code =1```
![1639114318(1)](https://user-images.githubusercontent.com/55161917/145522510-98a79060-ab6d-46a9-baf4-05d25ef8a9cb.png)
I think the line 19 should changed to  ``` buildResult.innerText = result != 0 ? 'Good' : 'Error'; ```

Yoda-Canada commented on src/api/status/src/views/builds.hbs:
```diff
@@ -32,6 +32,43 @@
+          </div>
+          <div
+            id="build-result"
+            class="col col-md-2 text-dark text-sm font-weight-bolder"
```
We can try to optimize the UI    
```    
     <div
            id="build-result"
            class="col col-md-2 text-dark text-sm font-weight-bolder fas fa-times-circle"
            style="color:red !important" 
          ></div>
 ```
![1639114960(1)](https://user-images.githubusercontent.com/55161917/145523296-25562b14-1d85-471c-b565-0d2fabfb6785.png)

mqnguyen5 commented:
I have made some adjustments based on @Kevan-Y and @Andrewnt219  reviews:

- Changed the `build-header` to use the [card layout](https://www.creative-tim.com/learning-lab/bootstrap/cards/material-dashboard), which helps adds extra padding to the left
- Used `card-link` class for the pusher's profile and the git SHA
- Removed the extra `buildHeaderTitle.innerText = 'Build Status';`

Andrewnt219 commented:
> I think the line 19 should changed to buildResult.innerText = result != 0 ? 'Good' : 'Error';

Yoda addressed this as well.

mqnguyen5 commented:
But inside the description for #2416, @humphd said:

> what the result code of the build was (e.g., 0 is "good" and non-zero is an "error")

Kevan-Y reviewed (changes requested):
Could you center this?
![image](https://user-images.githubusercontent.com/58233223/145686210-85f164c6-73c3-4358-93ed-8905d7e7a9bb.png)

Yoda-Canada commented:
@humphd  I just check the issue #2416, and your comment is that "what the result code of the build was  (e.g., 0 is "good" and non-zero is an "error")" ?  According to this comment, the code ` buildResult.innerText = result === 0 ? 'Good' : 'Error'; ` is correct. 
However, my understanding is that when the build is finished, the result code of the build is good.  I checked the data of "//dev.api.telescope.cdot.systems:4000/status".  The build had been done, but the code =1. 
![1639114318(1)](https://user-images.githubusercontent.com/55161917/145522510-98a79060-ab6d-46a9-baf4-05d25ef8a9cb.png)
I am confused that although the project has been completed, the result can also be an error, right?

humphd commented:
https://en.wikipedia.org/wiki/Exit_status

In Unix, a zero exit code means it worked, anything else is an error.

Yoda-Canada commented:
@mqnguyen5 your code ` buildResult.innerText = result != 0 ? 'Good' : 'Error'; ` is correct. My point of view is wrong, sorry for that.

humphd commented:
Good to clarify, there are lots of moving pieces here.

Yoda-Canada commented on src/api/status/public/js/build-log/build-header.js:
```diff
@@ -0,0 +1,38 @@
+const buildDuration = document.getElementById('build-duration');
+
+function renderBuildInfo({ githubData, startedAt, stoppedAt, result }) {
+  buildHeaderTitle.innerHTML = '';
```
Line 12 has changed, but it is different from Andrent219's comment.
![image](https://user-images.githubusercontent.com/55161917/145731065-ef97deba-e9e5-4434-9fd8-d7ecb5cde486.png)

Andrewnt219 reviewed (commented):
LGTM. Can you rebase and revert the change to `autodeploymentUrl`?

humphd commented on src/api/status/public/js/build-log/build-header.js:
```diff
@@ -0,0 +1,38 @@
+
+function renderBuildInfo({ githubData, startedAt, stoppedAt, result }) {
+  buildHeaderTitle.innerHTML = '';
+  buildSender.setAttribute('href', githubData.sender.html_url);
```
You don't need to use `setAttribute`, use the setter/getter on the DOM element API: `buildSender.href = ...`

humphd commented on src/api/status/public/js/build-log/build-header.js:
```diff
@@ -0,0 +1,38 @@
+  buildSender.setAttribute('href', githubData.sender.html_url);
+  buildSenderName.innerText = githubData.sender.login;
+  buildSenderImg.src = githubData.sender.avatar_url;
+  buildGitSHA.setAttribute('href', githubData.compare);
```
Same thing here

humphd reviewed (changes requested):
Two small things, looks good otherwise.

Fix and rebase please.

Kevan-Y reviewed (changes requested):
Not sure if it looks good this way
![image](https://user-images.githubusercontent.com/58233223/145752190-a7b6e289-bccd-4c37-b2a5-85b9af748948.png)
Maybe you can hide the part below on loading

You can element `hidden` of `id="build-header-info`
and have a quick check if  elem `build-header-info` is hidden or not after line 12 in `src/api/status/public/js/build-log/build-header.js` 
```js
if (buildHeaderInfo.hidden) {
    buildHeaderInfo.removeAttribute('hidden');
}
  ```

mqnguyen5 commented:
Changes have been made in response to @humphd and @Kevan-Y reviews. Sorry for not able to complete this PR sooner, I should've schedule my time better.

Andrewnt219 reviewed (approved):
Looks great. Well done 👍.

Yoda-Canada reviewed (approved):
Great！You've really done hard work.

Kevan-Y reviewed (approved):
LGTM!
